File: data/derived_cot/rq1_traces/translation/qwen/cpp_to_py/code_output/NumericEntityUnescaper.py
class NumericEntityUnescaper:
    @staticmethod
    def is_hex_char(c):
        c_lower = c.lower()
        return c.isdigit() or ('a' <= c_lower <= 'f')
    
    def replace(self, input_str):
        if not input_str:
            return ""
        out_list = []
        n = len(input_str)
        i = 0
        
        while i < n - 2:
            if input_str[i] == '&' and input_str[i+1] == '#':
                start = i + 2
                is_hex = False
                
                if start < n and input_str[start].lower() == 'x':
                    start += 1
                    is_hex = True
                
                j = start
                while j < n and self.is_hex_char(input_str[j]):
                    j += 1
                
                if j < n and input_str[j] == ';':
                    try:
                        num_str = input_str[start:j]
                        if is_hex:
                            entity_value = int(num_str, 16)
                        else:
                            entity_value = int(num_str, 10)
                        char_to_append = chr(entity_value)
                        out_list.append(char_to_append)
                        i = j + 1
                        continue
                    except ValueError:
                        pass
            
            out_list.append(input_str[i])
            i += 1
        
        out_list.append(input_str[i:])
        return ''.join(out_list)

File: data/derived_cot/rq1_traces/translation/qwen/py_to_cpp/code_output/NumericEntityUnescaper.cpp
#include <string>
#include <vector>
#include <cctype>
#include <stdexcept>
#include <locale>
#include <algorithm>

class NumericEntityUnescaper {
public:
    std::string replace(const std::string& input) {
        std::string result;
        size_t pos = 0;
        size_t len = input.length();

        while (pos < len - 2) {
            if (input[pos] == '&' && input[pos+1] == '#') {
                size_t start = pos + 2;
                bool is_hex = false;

                if (start < len) {
                    char c = input[start];
                    if (c == 'x' || c == 'X') {
                        start++;
                        is_hex = true;
                    }
                }

                if (start >= len) {
                    return result;
                }

                size_t end = start;
                while (end < len && NumericEntityUnescaper::is_hex_char(input[end])) {
                    end++;
                }

                if (end < len && input[end] == ';') {
                    try {
                        std::string substr = input.substr(start, end - start);
                        int entity_value = std::stoi(substr, nullptr, is_hex ? 16 : 10);
                        result += static_cast<char>(entity_value);
                        pos = end + 1;
                        continue;
                    } catch (...) {
                        return result;
                    }
                }
            }

            result += input[pos];
            pos++;
        }

        return result;
    }

    static bool is_hex_char(char c) {
        if (std::isdigit(c)) {
            return true;
        }
        char lower_c = std::tolower(c);
        return lower_c >= 'a' && lower_c <= 'f';
    }
};